### PR TITLE
Fix privacy leak in dashboard/facts/stats - Refs #1582

### PR DESCRIPTION
**Pull request by GregSutcliffe**
This applies a new search scope (based on Host.my_hosts) to FactValue. :my_facts and :my_hosts are used to restrict the data shown in the Dashboard, Facts, and Statistics tab.

I'd like to write some tests for this so we can ensure we don't break this later, but I've no idea where to start, as there doesn't appear to be any existing tests on the display of the these pages. How would I go about writing that?
#1582 also talks about classes, but I have a separate plan for that, so this effectively closes #1582

### CODE REVIEW

**Comment by bgupta** [comment]
I tested commit 3046400 (privacy leaks) yesterday, and it worked as advertised in my environment. -Brian

File: app/controllers/dashboard_controller.rb [hunk 1]
```diff
@@ -13,7 +13,8 @@ def index
 
   private
   def prefetch_data
-    @hosts = Host.search_for(params[:search])
+    my_hosts = User.current.admin? ? Host : Host.my_hosts
```
Comment (ohadlevy) [inline app/controllers/dashboard_controller.rb]: would it make more sense to change the my_hosts to include the admin test

File: app/controllers/fact_values_controller.rb [hunk 2]
```diff
@@ -11,7 +11,9 @@ class FactValuesController < ApplicationController
 
   def index
     begin
-      values = FactValue.no_timestamp_facts.search_for(params[:search],:order => params[:order])
+      # restrict allowed facts list based on the user permissions
+      my_facts = User.current.admin? ? FactValue : FactValue.my_facts
```
Comment (ohadlevy) [inline app/controllers/fact_values_controller.rb]: same here, this way we could start using my_facts and my_hosts regardless and ensure they are always scoped.

**Comment by ohadlevy** [comment]
i'm a bit not clear if I should merge the whole thing, as there seems to be commits for different topics..?

File: app/controllers/fact_values_controller.rb (hunk 2 above)
Comment (GregSutcliffe) [inline app/controllers/fact_values_controller.rb]: That would work I guess. I'll see if can make that work

**Comment by GregSutcliffe** [comment]
Ah crap, pushed my archlinux commit by accident. Let me fix that

**Comment by GregSutcliffe** [comment]
Ok, that's cleaner. I should pay more attention

File: app/controllers/fact_values_controller.rb (hunk 2 above)
Comment (ohadlevy) [inline app/controllers/fact_values_controller.rb]: ok, I'll wait for that patch before merging then ;)

**Comment by GregSutcliffe** [comment]
Ok, refactored for your reading pleasure :)

File: app/models/fact_value.rb [hunk 3]
```diff
@@ -13,6 +13,11 @@ class FactValue < Puppet::Rails::FactValue
   scope :timestamp_facts, :joins => [:fact_name],
               :conditions => ["fact_names.name = ?",:_timestamp]
 
+  scope :my_facts, lambda {
+    {:conditions => sanitize_sql_for_conditions(
+      [" (fact_values.host_id in (?))",fhs = (Host.my_hosts.map(&:id))])}
```
Comment (ohadlevy) [inline app/models/fact_value.rb]: whats fhs?

Comment (GregSutcliffe) [inline app/models/fact_value.rb]: I copied the code from the :my_hosts lambda function, which has:

app/models/host.rb: line 73
    domain_conditions    = sanitize_sql_for_conditions([" (hosts.domain_id in (?))",dms = (user.domains).map(&:id)])

There I assumed dms was some kind of temporary variable, maybe meaning "domains"? So I did something similar. Only now do I realise it's not re-used like dms is, so I guess I can update that.

File: app/models/report.rb [hunk 4]
```diff
@@ -24,15 +24,10 @@ class Report < ActiveRecord::Base
 
   # returns reports for hosts in the User's filter set
   scope :my_reports, lambda {
-    user = User.current
-    if user.admin?
-      conditions = { }
```
Comment (ohadlevy) [inline app/models/report.rb:29]: why did you remove the admin checks from here?

Comment (GregSutcliffe) [inline app/models/report.rb:29]: :my_reports relies on :my_hosts, which now has the Admin check internally.

Comment (GregSutcliffe) [inline app/models/report.rb:29]: I suppose we could leave it in if it's faster than running the rest of the lambda?

Comment (ohadlevy) [inline app/models/report.rb:29]: I guess it does, as it would simplify the SQL query (it would also use indexes better etc)

it would be select \* from reports vs select \* from reports where host_id in (all ids of all hosts)

Comment (GregSutcliffe) [inline app/models/report.rb:29]: Ok, i'll update. I can do the same in :my_facts too

**Comment by GregSutcliffe** [comment]
Updated with in-scope Admin checks for each model.

EDIT: Also rebased from latest develop, just to keep up to date.

**Comment by ohadlevy** [comment]
merged
